### ISSUE TITLE
Expiring TODO/FIXMEs

### ISSUE THREAD

**Issue by msanders** [realm/SwiftLint#727]
It would be cool if there were a way to add expiring TODOs with SwiftLint.

I.e. `// TODO [2016-06-01]`, which would only warn after the given date.

Right now there isn't a great way of making use of TODOs with SwiftLint without either turning the warning off or adding noise to the project.

**Comment by yonaskolb** [comment]
This is a great idea! You could remind yourself to remove some outdated API when a new version of iOS comes out for example

**Comment by jpsim** [comment]
Great idea!

**Comment by jeremy-w** [comment]
Unless things have changed, Xcode only picks up `TODO:` as a group, so the proposed syntax would stop it from surfacing it in the method list popup. I recommend `[before: 2016-06-01] TODO: Something` as the format.

I use a variation of [Google's TODO comment recommendation](https://google.github.io/styleguide/cppguide.html#TODO_Comments) that works around this by prefixing the author info, like so:

```
(jeremy-w/2016-08-03)TODO: Maybe this should be FIXME?
```

(I stamp it with the current date so I get an implicit "oh that must not have been terribly urgent" read on a comment without having to hit up version control history to track when the TODO landed in the source.)

I handle this simple formatting with a custom rule:

```yaml
    todo_format:
        match_kinds: comment
        message: >
            Prefix TODO/FIXME/etc with email prefix and date,
            so if you're <name@example.com>, then you'd write:
            // (name/2016-07-28)TODO: Document custom rule
        # This isn't a perfect check, but it should guide folks the right
        # direction for common cases.
        regex: "[^)](TODO|FIXME|XXX|HACK|!{3}|[?]{3}):"
```

(That should probably be a non-capturing group, come to think.)

**Comment by AliSoftware** [comment]
That's not a solution to the problem, but for TODOs, we don't use SwiftLint and instead use a separate Build Phase which `grep` the files for `(TODO|FIXME|\!\!\!|\?\?\?):` and generate warnings from that.

The main benefit in having a separate Build Phase for that is that in fact we can execute that `grep` only in Analyze builds, not in standard builds. This means that we don't see any TODO warnings when we "Build" everyday, and only see them when we do an "Analyze". This avoids warning pollution.

Here's our the build phase we use, for reference:
```
if [ "${RUN_CLANG_STATIC_ANALYZER}" == "YES" ]; then
  echo "Printing TODOs as warnings..."
  KEYWORDS="TODO:|FIXME:|\?\?\?:|\!\!\!:"
  FILE_EXTENSIONS="swift|h|m|mm|c|cpp"
  find -E "${SRCROOT}" -ipath "${SRCROOT}/Carthage" -o -ipath "${SRCROOT}/Pods" -prune -o \( -regex ".*\.($FILE_EXTENSIONS)$" \) -print0 | xargs -0 egrep --with-filename --line-number --only-matching "($KEYWORDS).*\$" | perl -p -e "s/($KEYWORDS)/ warning: \$1/"
fi
```

Sure that solution is not the same as this "Expiring TODO" features presented in this issue (which would be quite cool as well) but I figure at least it could make your life easier and avoid some pollution on everyday build cycles, so HTH!

**Comment by namolnad** [comment]
@msanders did you every figure out some solution for this? seems it may have lost momentum here, so curious if you created a private rule or anything like that

**Comment by sindresorhus** [comment]
For inspiration, here's my take on it (in JavaScript land): https://github.com/sindresorhus/eslint-plugin-unicorn/blob/master/docs/rules/expiring-todo-comments.md

**Comment by namolnad** [comment]
@sindresorhus that one comprehensive and powerful rule! I've started out by implementing it for a date only, but you've definitely provided inspiration!

**Comment by namolnad** [comment]
@marcelofabri I think we should be able to close this now due to #2911 🎉
